### PR TITLE
simplify backend storage registry HTTP client

### PR DESCRIPTION
**Pull request by changweige**
1. No more conversion and judgment on registry response status code in the underlying connection layer. Just let the upper layer process the status code.
2. Let the registry client only use the helper function `request` rather than calling the underlying `call` method.
3. Simplify how to construct the URL to access blobs

`Connection` layer is a low-level transparent layer used by many modules, it is hard to say what status code should be treated as Success in such a lower layer. Let the upper layer handle the status code itself. Connection Ok(()) only means the request is successfully sent and processed by the registry server side.

The current registry client wrongly treats some responses as success and copies the response body to upper layer as image data chunks.

```
[2022-12-12 17:59:24.056751 +08:00] ERROR [/home/gechangwei/.cargo/git/checkouts/fuse-backend-rs-891780d38d588b3b/f717974/src/api/server/sync_io.rs:1180] fuse: reply error header OutHeader { len: 16, error: -5, unique: 700 }, error Os { code: 5, kind: Uncategorized, message: "Input/output error" }
[2022-12-12 17:59:26.057091 +08:00] WARN [storage/src/cache/state/blob_state_map.rs:120] Waiting for backend IO expires. chunk index 217, compressed offset 4329908
[2022-12-12 17:59:26.059065 +08:00] ERROR [error/src/error.rs:21] Error:
	"request for 416877 bytes but got 346 bytes"
	at storage/src/cache/mod.rs:209
	note: enable `RUST_BACKTRACE=1` env to display a backtrace
[2022-12-12 17:59:26.059094 +08:00] ERROR [/home/gechangwei/.cargo/git/checkouts/fuse-backend-rs-891780d38d588b3b/f717974/src/api/server/sync_io.rs:1180] fuse: reply error header OutHeader { len: 16, error: -5, unique: 702 }, error Os { code: 5, kind: Uncategorized, message: "Input/output error" }
[2022-12-12 17:59:28.059243 +08:00] WARN [storage/src/cache/state/blob_state_map.rs:120] Waiting for backend IO expires. chunk index 217, compressed offset 4329908
[2022-12-12 17:59:28.061074 +08:00] ERROR [error/src/error.rs:21] Error:
	"request for 416877 bytes but got 346 bytes"
	at storage/src/cache/mod.rs:209


	"request for 1208223 bytes but got 346 bytes"
	"request for 2028282 bytes but got 346 bytes"
	"request for 2176 bytes but got 346 bytes"
	"request for 1456899 bytes but got 346 bytes"
	"request for 2914838 bytes but got 346 bytes"
	"request for 2487163 bytes but got 346 bytes"
	"request for 2989585 bytes but got 346 bytes"
	"request for 1875970 bytes but got 346 bytes"
	"request for 3051237 bytes but got 346 bytes"
	"request for 2301065 bytes but got 346 bytes"
	"request for 2103970 bytes but got 346 bytes"
	"request for 2986356 bytes but got 346 bytes"
	"request for 1813040 bytes but got 346 bytes"
	"request for 2179829 bytes but got 346 bytes"
	"request for 2192490 bytes but got 346 bytes"
	"request for 2579762 bytes but got 346 bytes"
	"request for 2457311 bytes but got 346 bytes"
	"request for 2220056 bytes but got 346 bytes"
	"request for 290185 bytes but got 346 bytes"
	"request for 2372729 bytes but got 346 bytes"
	"request for 692 bytes but got 346 bytes"

```

In addition, nydusd storage backend only serves read-only requests, so remove unnecessary request body to reduce number of function
parameters.

### CODE REVIEW

File: storage/src/backend/registry.rs [hunk 1]
```diff
@@ -438,17 +433,16 @@ impl RegistryReader {
...
                     );
 
-                    // Try to request registry server with `authorization` header again
-                    let resp = self
+                    // Cache authorization header for next request
+                    // Cache the auth earlier, so other threads have bigger chance to reuse the auth.
+                    self.state.cached_auth.set(auth_header.as_str());
+
+                    // Try to request registry server with `authorization` header again,
+                    // It's exactly a resend of the original request!
```
Comment (imeoer) [inline storage/src/backend/registry.rs:441]: It seems `resend` can't be used as noun.

File: storage/src/backend/registry.rs [hunk 2]
```diff
@@ -256,7 +254,6 @@ impl RegistryState {
                 Some(ReqBody::Form(form)),
                 &mut headers,
                 true,
-                true,
             )
             .map_err(|e| einval!(format!("registry auth server request failed {:?}", e)))?;
         let ret: TokenResponse = token_resp.json().map_err(|e| {
```
Comment (imeoer) [inline storage/src/backend/registry.rs:259]: Where is the `token_resp.status()` checked?

File: storage/src/backend/object_storage.rs [hunk 3]
```diff
@@ -75,15 +76,7 @@ where
...
-                url.as_str(),
-                None,
-                None,
-                &mut headers,
-                true,
-                false,
-            )
+            .call::<&[u8]>(Method::HEAD, url.as_str(), None, None, &mut headers, false)
             .map_err(ObjectStorageError::Request)?;
         let content_length = resp
```
Comment (imeoer) [inline storage/src/backend/object_storage.rs:81]: Where is the `resp.status()` checked?

File: storage/src/backend/object_storage.rs [hunk 4]
```diff
@@ -116,10 +117,19 @@ where
...
             .map_err(ObjectStorageError::Request)?;
-        Ok(resp
-            .copy_to(&mut buf)
-            .map_err(ObjectStorageError::Transport)
-            .map(|size| size as usize)?)
+
+        let status = resp.status();
+
+        if status == StatusCode::PARTIAL_CONTENT || status == StatusCode::OK {
+            // TODO: Support multiple range request
```
[the remark this answers is not in the document]
Comment (changweige) [reply storage/src/backend/object_storage.rs:131]: The comment line is removed.

File: storage/src/backend/registry.rs (hunk 1 above)
Comment (changweige) [reply storage/src/backend/registry.rs]: fixed

File: storage/src/backend/registry.rs (hunk 2 above)
Comment (changweige) [reply storage/src/backend/registry.rs:267]: Thanks for pointing it out. It is handled now

File: storage/src/backend/object_storage.rs (hunk 3 above)
Comment (changweige) [reply storage/src/backend/object_storage.rs:88]: fixed

File: storage/src/backend/object_storage.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (imeoer) [reply storage/src/backend/object_storage.rs:131]: It seems still here. others LGTM!

**Review by imeoer — APPROVED** [review]

**Review by jiangliu — APPROVED** [review]

**Comment by jiangliu** [comment]
Please help to solve the conflicts:)

**Comment by changweige** [comment]
It's not easy to solve the conflicts, so I will try to make up another PR

**Pull request closed without merging**